src/Apps/ArtAdvisor/06-Discovery-Neartext/weaviate-db.ts

```diff
@@ -0,0 +1,264 @@
+       * same artwork class that is being supplied here.
+       */
+      userClass?: string
+    } = {}
+  ) {
+    const url = options.url || process.env.WEAVIATE_URL
+    if (!url) throw new Error("Please provide url or set WEAVIATE_URL")
+
+    this.client = weaviate.client({ host: url })
+    this.artworkClass = options.artworkClass || DEFAULT_ARTWORKS_CLASS
+    this.userClass = options.userClass || DEFAULT_USERS_CLASS
+  }
+
+  async getUsers() {
+    const response = await this.client.graphql
+      .get()
+      .withClassName(this.userClass)
+      .withFields("name _additional { id }")
+      .do()
+
+    const users = response.data.Get.DiscoveryUsers.map(user => {
+      return {
+        id: user._additional.id,
+        name: user.name,
+      }
+    })
+
+    return users
+  }
+
+  /**
+   * Fetch a user
+   */
+  async getUser({
+    internalID,
+    id,
+  }: {
+    /** Mongo ID for Gravity User */
+    internalID?: string
+    /** Weaviate UUID for DiscoveryUser */
+    id?: string
+  }): Promise<DiscoveryUser> {
+    if (!internalID && !id)
+      throw new Error("Must provide either internalID or id (UUID)")
+
+    if (internalID) throw new Error("Not implemented yet")
+
+    if (!id) {
+      // temporary, while using test users
+      throw new Error("Missing id")
+    }
+
+    const response = await this.client.graphql
+      .get()
+      .withClassName(this.userClass)
+      .withWhere({
+        path: ["id"],
+        operator: "Equal",
+        valueString: id,
+      })
+      .withFields(
+        `
+        name
+        likedArtworks {
+          ... on DiscoveryArtworks { internalID }
+        }
+        dislikedArtworks {
+          ... on DiscoveryArtworks { internalID }
+        }
+        `
+      )
+      .do()
+
+    const {
+      name,
+      likedArtworks,
+      dislikedArtworks,
+    } = response.data.Get.DiscoveryUsers[0]
+
+    return {
+      id,
+      // internalID, // TODO: once we create real Gravity users
+      name,
+      likedArtworkIds: (likedArtworks || []).map(w => w.internalID),
+      dislikedArtworkIds: (dislikedArtworks || []).map(w => w.internalID),
+    } as DiscoveryUser
+  }
+
+  /**
+   * Create a liked or disliked artwork for a user
+   *
+   * @returns true on success, false if user already liked or disliked artwork
+   */
+  async reactToArtwork({
+    userId,
+    userInternalID,
+    artworkInternalID,
+    reaction,
+  }: {
+    userId?: UUID
+    userInternalID?: MongoID
+    artworkInternalID: MongoID
+    reaction: "like" | "dislike"
+  }) {
+    if (userInternalID) {
+      userId = generateUuid5(userInternalID)
+    }
+    if (!userId)
+      throw new Error("Provide either userId (UUID) or userInternalID")
+
+    // bail if user already liked or disliked artwork
+
+    const user = await this.getUser({ id: userId })
+    if (user.likedArtworkIds.includes(artworkInternalID)) return false
+    if (user.dislikedArtworkIds.includes(artworkInternalID)) return false
```

Comment:
This was interesting — Weaviate will happily create multiple refs for the same artwork; will even let you add the same artwork to both likes _and_ dislikes. So this short circuits to try and avoid having an artwork getting liked/disliked more than one time.